exec/vector/src/main/java/org/apache/drill/exec/record/MaterializedField.java

```diff
@@ -49,39 +54,79 @@ private MaterializedField(String name, MajorType type, LinkedHashSet<Materialize
     this.children = children;
   }
 
+  private MaterializedField(String name, MajorType type, int size) {
+    this(name, type, new LinkedHashSet<>(size));
+  }
+
+  private <T> void copyFrom(Collection<T> source, Function<T, MaterializedField> transformation) {
+    Preconditions.checkState(children.isEmpty());
+    source.forEach(child -> children.add(transformation.apply(child)));
+  }
+
+  public static MaterializedField create(String name, MajorType type) {
+    return new MaterializedField(name, type, 0);
+  }
+
   public static MaterializedField create(SerializedField serField) {
-    LinkedHashSet<MaterializedField> children = new LinkedHashSet<>();
-    for (SerializedField sf : serField.getChildList()) {
-      children.add(MaterializedField.create(sf));
+    MaterializedField field = new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), serField.getChildCount());
+    if (OFFSETS_FIELD.equals(field)) {
+      return OFFSETS_FIELD;
     }
-    return new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), children);
+    field.copyFrom(serField.getChildList(), MaterializedField::create);
+    return field;
   }
 
-  /**
-   * Create and return a serialized field based on the current state.
-   */
-  public SerializedField getSerializedField() {
-    SerializedField.Builder serializedFieldBuilder = getAsBuilder();
-    for(MaterializedField childMaterializedField : getChildren()) {
-      serializedFieldBuilder.addChild(childMaterializedField.getSerializedField());
+  public MaterializedField copy() {
+    return copy(getName(), getType());
+  }
+
+  public MaterializedField copy(MajorType type) {
+    return copy(name, type);
+  }
+
+  public MaterializedField copy(String name) {
+    return copy(name, getType());
+  }
+
+  public MaterializedField copy(String name, final MajorType type) {
+    if (this == OFFSETS_FIELD) {
+      return this;
     }
-    return serializedFieldBuilder.build();
+    MaterializedField field = new MaterializedField(name, type, getChildren().size());
+    field.copyFrom(getChildren(), MaterializedField::copy);
```

Comment:
The copy always copies children, which, as we saw in another PR, is a bit murky.

1. Is this ever used for maps? If not, then whether we copy children is ambiguous. (Doesn't matter if we copy children for vectors without them.)
2. If used maps, will copy the children. But, when we create that map clone, we'll add the child vectors one by one, which will create the child fields. Will we end up with duplicates?

Since we are touching this area, should we clean it up?

Reply:
`children` in `MaterializedField` is a `Set`. Please see `equals()` and `hash()` for `MaterializedField`.

Reply:
My point is not how things are implemented. The point is: we almost never want to copy children when copying a `MaterializedField`. Why? Because the consumer of that copy may be code that creates a new vector. When it does, it will add child vectors (copies of the source vectors), and that action will add new child fields.

This is why I said that "copy" is misleading: we need to understand the context in which we make the copy, and possibly name the method accordingly: "copyForNewVector", "fullCopy", "copyIfNeeded", and so on. All of these should have comments that explain the use case that they serve.

Sorry that this is so complex; it is just the way the code has evolved. As I recently found, it is quite hard to change entrenched code behavior, even when it is not quite right.

Reply:
There is a case in `UnnestRecordBatch` where we have to keep a copy of Materialized field along with it's children of incoming repeated type vector which may be repeated map as well. This is to detect if there is any schema change with new incoming batch as compared to previous one.

Also in case of creating a Map vector in output container one can call `container.addOrGet(field, null) ` if the field contains children information then this call will take care of adding all the children vectors in Map vector as well. So you don't have to explicitly add child vectors.

Reply:
@paul-rogers In cases where `deep` copy is not necessary, `MaterializedField` already provides `create()` method that takes name and type. There is no need to introduce `copyForNewVector`, `fullCopy` or `copyIfNeeded` IMO.